### PR TITLE
Wire network interface into libpod

### PR DESCRIPTION
**Pull request by Luap99**
### Wire network interface into libpod

Make use of the new network interface in libpod.

This commit contains several breaking changes:
- podman network create only outputs the new network name and not file
  path.
- podman network ls shows the network driver instead of the cni version
  and plugins.
- podman network inspect outputs the new network struct and not the cni
  conflist.
- The bindings and libpod api endpoints have been changed to use the new
  network structure.

The container network status is stored in a new field in the state. The
status should be received with the new `c.getNetworkStatus`. This will
migrate the old status to the new format. Therefore old containers should
continue to work correctly in all cases even when network connect/
disconnect is used.

New features:
- podman network reload keeps the ip and mac for more than one network.
- podman container restore keeps the ip and mac for more than one
  network.
- The network create compat endpoint can now use more than one ipam
  config.

The man pages and the swagger doc are updated to reflect the latest
changes.

### Drop OCICNI dependency

We do not use the ocicni code anymore so let's get rid of it. Only the
port struct is used but we can copy this into libpod network types so
we can debloat the binary.

The next step is to remove the OCICNI port mapping form the container
config and use the better PortMapping struct everywhere.

### CODE REVIEW

**Comment by edsantiago** [comment]
There are no such `skip` helpers, because I consider such usage (checking for OS/distro) improper: the correct way is to test for functionality. ISTM that you actually have that?
```bash
    cni=/usr/libexec/cni/bridge       ! FIXME: there should be a better way to find this, but `podman info` does not help
    test -x $cni || skip "CNI executable $cni not installed"
    run $cni --version
    cni_version=$(expr "$output" : ".* plugin version \(.*\)")
    [[ $cni_version -ge 1.0 ]] || skip "Test requires CNI >= 1.0 (found: $cni_version)"
```
Using `[[` will treat `unknown` as `< 1.0` without emitting a warning. Using `test` will also work, but with "integer expression expected" warning.

**Comment by Luap99** [comment]
`/usr/libexec/cni/bridge` always prints unknown even for 1.0 so this does not work.

**Comment by Luap99** [comment]
Also it looks like the path is `/usr/lib/cni/bridge` on debian/ubuntu

**Comment by edsantiago** [comment]
> `/usr/libexec/cni/bridge` always prints unknown even for 1.0 so this does not work.

Well, that's obviously a bug and it must be fixed; but equally obviously, it can't get fixed in time for your PR.

What is the shortest, simplest way to behavior-check and identify good-new-cni vs old-bad-cni? Is there a `podman network foo bar` that will exit error with old-cni and success on new-cni? If so, you could add such a test to `setup()` in the system test file that requires it. (If multiple test files, then it's better to add `require_cni_with_FEATURENAME` to `helpers.bash`).

**Comment by Luap99** [comment]
For now this is the only test which requires the new functionality so it doesn't make sense to use the same test to check if the test should be skipped so I will remove the parts which require v1.0. In future PRs I will add functionality to test this in a reasonable way.

**Comment by Luap99** [comment]
@mheon PTAL

File: libpod/runtime.go [hunk 1]
```diff
@@ -462,17 +462,20 @@ func makeRuntime(ctx context.Context, runtime *Runtime) (retErr error) {
 		}
 	}
 
-	// If we need to make a default network - do so now.
-	if err := defaultnet.Create(runtime.config.Network.DefaultNetwork, runtime.config.Network.DefaultSubnet, runtime.config.Network.NetworkConfigDir, runtime.config.Engine.StaticDir, runtime.config.Engine.MachineEnabled); err != nil {
```
Comment (mheon) [inline libpod/runtime.go:466]: This is done by the ContainerNetwork as it initializes now?

**Comment by mheon** [comment]
Preliminary LGTM. This is too big for an in-depth review, but I can confirm that the Libpod bits look good.

I think we have more work to do in the Libpod config (rewrite network config to better support v6), but this is already large and I really don't want to put any more code into it.

File: libpod/runtime.go (hunk 1 above)
Comment (Luap99) [reply libpod/runtime.go:466]: The default network is now in memory only. We won't write a cni config file. That allows the user to change the default network name/subnet in containers.conf and the changes will be picked up by the next podman command.

**Comment by Luap99** [comment]
I will rebase to pick up #11485, this contains useful upgrade tests for this.

**Comment by Luap99** [comment]
@TomSweeneyRedHat PTAL at the man pages

**Comment by Luap99** [comment]
And my new network upgrade tests failed. I guess it was a good idea to write them.

**Comment by TomSweeneyRedHat** [comment]
Manpage changes LGTM.
I did a very high-level skim review and it LGTM and we've happy green test buttons.

**Comment by Luap99** [comment]
/hold cancel
we have a 3.4 branch so this can merge

**Comment by Luap99** [comment]
@mheon @baude PTAL

**Comment by mheon** [comment]
You have a vendor conflict. LGTM once that's resolved.

**Pull request merged**